File: Spiral_Matrix.cpp
/*54. Spiral Matrix  QuestionEditorial Solution  My Submissions
Total Accepted: 74909 Total Submissions: 312809 Difficulty: Medium
Given a matrix of m x n elements (m rows, n columns), return all elements of the matrix in spiral order.

For example,
Given the following matrix:

[
 [ 1, 2, 3 ],
 [ 4, 5, 6 ],
 [ 7, 8, 9 ]
]
You should return [1,2,3,6,9,8,7,4,5].

Hide Company Tags Microsoft Google Uber
Hide Tags Array
Hide Similar Problems (M) Spiral Matrix II
*/



/*
traverse right and increment right
then traverse down and decrement down
then traverse left and decrement left
then traverse up and increment up

when traverse left or up, need to check whether the row or col still exists to prevent duplicates*/

/////////////////////////////////////////////////////////////////////////////////////
//C++
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> ret;
        if(matrix.size()==0 || matrix[0].size()==0) return ret;
        int right = 0;
        int down = 0;
        int left = matrix[0].size()-1;
        int up = matrix.size()-1;
        
        while(down<=up && right<=left){
            //traverse right;
            for(int i = right; i <= left; ++i){
                ret.push_back(matrix[down][i]);
            }
            down++;
            
            //traverse down
            for(int i = down; i <= up; ++i){
                ret.push_back(matrix[i][left]);
            }
            left--;
            
            if(down<=up){
                //traverse left
                for(int i = left; i >= right; --i){
                    ret.push_back(matrix[up][i]);
                }
            }
            up--;
            
            if(right<=left){
                //traverse up
                for(int i = up; i >= down; --i){
                    ret.push_back(matrix[i][right]);
                }
            }
            right++;
        }
        return ret;
    }
};




/////////////////////////////////////////////////////////////////////////////////////
//Java
public class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> ret = new ArrayList<>();
        if(matrix.length==0 || matrix[0].length==0) return ret;
        int right = 0;
        int down = 0;
        int left = matrix[0].length-1;
        int up = matrix.length-1;
        
        while(right<=left && down<=up){
            //traverse right
            for(int i = right; i <= left; ++i){
                ret.add(matrix[down][i]);
            }
            down++;
            
            //traverse down
            for(int i = down; i <= up; ++i){
                ret.add(matrix[i][left]);
            }
            left--;
            
            //traverse left
            if(down<=up){//check down and up
                for(int i = left; i >= right; --i){
                    ret.add(matrix[up][i]);
                }
            }
            up--;
            
            //traverse up
            if(right<=left){//check right and left
                for(int i = up; i >= down; --i){
                    ret.add(matrix[i][right]);
                }
            }
            right++;
        }
        return ret;
    }
}

File: Spiral_Matrix.swift
/*
Given an m x n matrix, return all elements of the matrix in spiral order.

 

Example 1:


Input: matrix = [[1,2,3],[4,5,6],[7,8,9]]
Output: [1,2,3,6,9,8,7,4,5]
Example 2:


Input: matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]
Output: [1,2,3,4,8,12,11,10,9,5,6,7]
 

Constraints:

m == matrix.length
n == matrix[i].length
1 <= m, n <= 10
-100 <= matrix[i][j] <= 100

Hint 1
Well for some problems, the best way really is to come up with some algorithms for simulation. Basically, you need to simulate what the problem asks us to do.

Hint 2
We go boundary by boundary and move inwards. That is the essential operation. First row, last column, last row, first column and then we move inwards by 1 and then repeat. That's all, that is all the simulation that we need.

Hint 3
Think about when you want to switch the progress on one of the indexes. If you progress on
i
out of
[i, j]
, you'd be shifting in the same column. Similarly, by changing values for
j
, you'd be shifting in the same row. Also, keep track of the end of a boundary so that you can move inwards and then keep repeating. It's always best to run the simulation on edge cases like a single column or a single row to see if anything breaks or not.
*/

/*
Solution 1
check layer by layer

1. [r1, c1] ... [r1, c2]
2. [r1+1, c2] ... [r2, c2]
3. [r2, c2-1] ... [r2, c1]
4. [r2-1, c1] ... [r1+1, c1]

Time Complexity: O(mn)
Space Complexity: O(1)
*/
class Solution {
    func spiralOrder(_ matrix: [[Int]]) -> [Int] {
        guard !matrix.isEmpty, !matrix[0].isEmpty else { return [] }
        let m = matrix.count
        let n = matrix[0].count
        
        var r1 = 0
        var r2 = m-1
        var c1 = 0
        var c2 = n-1
        
        var res = [Int]()
        while r1 <= r2, c1 <= c2 {
            res.append(contentsOf: Array(matrix[r1][c1...c2]))
            
            if r1 < r2 {
                for r in (r1+1)...r2 {
                    res.append(matrix[r][c2])
                }
            } 
            
            if r1 < r2, c1 < c2 {
                for c in stride(from: c2-1, through: c1, by: -1) {
                    res.append(matrix[r2][c])
                }

                for r in stride(from: r2-1, to: r1, by: -1) {
                    res.append(matrix[r][c1])
                }
            }
            
            r1 += 1
            r2 -= 1
            c1 += 1
            c2 -= 1
        }
        
        return res
    }
}